EIP-3675: Execution engine must validate length of the block's extraData (#4351)
ethereum/EIPs: EIPS/eip-3675.md

PoW blocks that are descendants of any terminal PoW block **MUST NOT** be imported. This implies that a terminal PoW block will be the last PoW block in the canonical chain.

### Constants

| Name | Value |
|-|-|

> | **`MAX_BYTES_PER_EXTRA_DATA`** | `32` |

Reviewer: ```suggestion
| **`MAX_EXTRA_DATA_BYTES`** | `32` |
```
Maybe update to this to be consistent with the consensus spec?

Reviewer 2: no need for "PER" here. I was mistaken. When it is the length of a single list in a block, we don't use "PER". As in `MAX_ATTESTATIONS`. "PER" is used more for conversions

Let's use @lightclient's suggestion here

> | **`MAX_BYTES_PER_EXTRA_DATA`** | `32` |

Reviewer: ```suggestion
| **`MAX_EXTRA_DATA_BYTES`** | `32` |
```
Maybe update to this to be consistent with the consensus spec?

> | **`MAX_BYTES_PER_EXTRA_DATA`** | `32` |

Reviewer: ```suggestion
| **`MAX_EXTRA_DATA_BYTES`** | `32` |
```
Maybe update to this to be consistent with the consensus spec?

| **`nonce`** | `0x0000000000000000` |  |
| **`ommers`** | `[]` | `RLP([]) = 0xc0`  |

> Beginning with `TRANSITION_BLOCK`, the validation of the block's **`extraData`** field changes: The length of the block's **`extraData`** **MUST** be less or equal to **`MAX_BYTES_PER_EXTRA_DATA`** bytes.

Reviewer: ```suggestion
Beginning with `TRANSITION_BLOCK`, the validation of the block's **`extraData`** field changes: The length of the block's **`extraData`** **MUST** be less or equal to **`MAX_EXTRA_DATA_BYTES`** bytes.
```
And here.

Reviewer 2: suggest "less than or" rather than "less or"

Author: ACK! Done in db1c213.

* Remove all validation rules that are evaluated over the list of ommers and each member of this list.
* Add verification of the fields noted in [block structure](#block-structure) as their specified constant value.
* Add verification of block validity with respect to the PoS consensus rules, i.e. assert that there **MUST** be a corresponding `POS_CONSENSUS_VALIDATED` event for the block to consider it valid.

> * Add verification of the block's **`extraData`** field. i.e. the length of the block's **`extraData`** **MUST** be less or equal to 32 bytes.

Reviewer 2: Suggest "32 bytes" be defined as a constant earlier in the spec. e.g. `MAX_BYTES_PER_EXTRA_DATA`

It is recommended for the client software to not propagate descendants of any terminal PoW block to reduce the load on processing the P2P component and stop operating in the environment with unknown security properties.

### Restricting the length of `extraData`

> The `extraData` field has been specified as unbounded in the PoW spec even though mainnet and most PoW testnets only utilize `32` bytes for it.  `extraData` fields of greater length are used by clique testnets and other networks to carry special signature/consensus schemes. This EIP restricts the length of `extraData` to `32` bytes since any network that is transitioning from another consensus mechanism to a beacon chain PoS consensus mechanism no longer needs extended or unbounded `extraData`.

Reviewer 2: The yellow paper (which defines ethereum pow/mainnet consensus) notes that "... extraData must be at most
32 bytes." Thus it is not unbounded for PoW network spec as noted in the original comment.

Suggest using the following text (or similar):

```suggestion
The `extraData` field is defined as a maximum of `32` bytes in the yellow paper. Thus mainnet and most PoW testnets cap the value at `32` bytes.  `extraData` fields of greater length are used by clique testnets and other networks to carry special signature/consensus schemes. This EIP restricts the length of `extraData` to `32` bytes because any network that is transitioning from another consensus mechanism to a beacon chain PoS consensus mechanism no longer needs extended or unbounded `extraData`.
```

Author: Thanks for the suggestion. Adopted in db1c213.
